### PR TITLE
feat: add FlyDSL debug/tiled programming skills to GEAK

### PR DESCRIPTION
**Pull request by amd-danli103**
## Summary

#155 added the initial FlyDSL integration in GEAK and introduced optimization-oriented support through `flydsl-optimization`.

This PR further adds two missing companion skills for earlier-stage FlyDSL workflows:
- `flydsl-tile-programming`: for kernel bring-up, structural kernel development, and Triton-to-FlyDSL porting
- `flydsl-debug-kernel`: for correctness, compilation, and stability debugging

Together, these skills make GEAK's FlyDSL support more complete across the full workflow.

## Why this is needed

In practice, many FlyDSL tasks are not ready for optimization first. The agent often needs to either build the first correct kernel structure or debug a runnable-but-broken kernel before performance work becomes meaningful.

These two skills fill that gap and give GEAK guidance for the stages that happen before `flydsl-optimization` becomes useful. GEAK needs explicit support to turn a runnable kernel into one that is structurally sound, correct, and stable. That is the foundation for any later work, including future performance-focused skills.

## How this differs from #155

#155 mainly focused on pipeline enablement and optimization once FlyDSL tasks could be recognized and run inside GEAK.

This PR covers two different scenarios that #155 did not address directly:
- building the initial FlyDSL kernel structure
- debugging correctness / compilation / stability issues before optimization

So this PR complements #155 rather than overlapping with it.

## Test coverage

This PR also updates `tests/skills/test_skill_runtime.py` to verify that the new FlyDSL skills are runtime-discoverable through `SkillRuntime`, rather than only being present as files in the repo.

### Skills discoverable during runtime
Run:

```bash
python3 -m pytest tests/skills/test_skill_runtime.py
```

### Agent loads skills correctly
Designed some test cases for validation. Will add more details into the comments.

#### flydsl-tile-programming
This skill should be used in the following situations:

- Writing a new FlyDSL `@flyc.kernel` from scratch.
- Porting an existing torch/Triton kernel to FlyDSL.
- Building a first-correct FlyDSL kernel before any performance tuning.
- Refactoring an existing FlyDSL kernel when the goal is to restructure its tiling, layout, partitioning, or launch organization.

Across these cases, the common pattern is that the task requires structural kernel design or refactoring.

#### flydsl-debug-kernel
- Debugging an existing FlyDSL kernel that already has a runnable entry point, repro command, or failing harness.
- Investigating numerical correctness failures, such as NaN/Inf outputs or mismatches in an existing FlyDSL kernel.
- Fixing compilation or frontend tracing failures in an existing FlyDSL kernel.
- Triaging runtime stability problems in an existing FlyDSL kernel.
- Investigating correctness or stability regressions introduced while optimizing an existing FlyDSL kernel.

Across these cases, the common pattern is that the kernel already exists and the task is to reproduce, localize, and fix a correctness, compilation, or stability problem.

### How skills benefit the tasks
Designed some test cases for validation. Will add more details into the comments.

### CODE REVIEW

**Comment by amd-danli103** [comment]
## Validation
### Agent loads skills correctly
#### Case 1. When a GEAK FlyDSL task needs structural kernel work, e.g. writing a new flyDSL kernel
```
geak --repo "$REPO" \
  --kernel-url "$REPO/base_softmax.py" \
  --task "用 FlyDSL 写一个 first-correct 的 softmax kernel，给最小 correctness harness，先不用做性能优化。" \
  --yolo
```
image

#### Conclusion: `flydsl-tile-programming/SKILL.md` successfully loaded and new kernel development done
We validated from the `~/GEAK/optimization_logs/softmax_20260423_071104/results/round_1/parallel_0/traj.json` log that the correct skill `flydsl-tile-programming` has been loaded.

#### Case 2. When a GEAK FlyDSL task needs debugging, e.g. a flyDSL kernel produces NaN/inf
```
geak --repo "$REPO" \
  --kernel-url "$REPO/buggy_softmax_kernel.py" \
  --test-command "python repro_buggy_softmax_nan.py" \
  --task "这是一个已有的FlyDSL softmax kernel。先运行 repro_buggy_softmax_nan.py 复现 NaN/inf 和 correctness mismatch，再定位 root cause 并修复。保持现有 kernel 入口不变，先不要做性能优化，也不要整体重写。" \
  --yolo
```
image

#### Conclusion: `flydsl-debug-kernel/SKILL.md` successfully loaded and the bug fix done
We validated from the `~/GEAK/optimization_logs/softmax_20260424_034547/results/round_1/parallel_0/traj.json` log that the correct skill `flydsl-debug-kernel` has been loaded.
- Before - NaN exists, nan_count = 2052
image

- After - NaN issue fixed, nan_count = 0
image

**Comment by amd-danli103** [comment]
### Under what circumstances would the skills be used (more examples)
#### flydsl-tile-programming
- Writing a new FlyDSL `@flyc.kernel` from scratch. (see Validation--Agent loads skills correctly--case1)
- Porting an existing torch/Triton kernel to FlyDSL.
```
geak --repo "$REPO" \
  --kernel-url "$REPO/triton_matrix_transpose.py" \
  --test-command "python triton_matrix_transpose.py --rows 257 --cols 513 --seed 1234" \
  --task "Port this Triton 2D matrix transpose kernel to a first-correct FlyDSL kernel. Preserve transpose semantics and the current validation path. Focus on tile/layout/partition/launch structure only. Do not optimize performance or change the operator." \
  --yolo
```
image

#### flydsl-debug-kernel
- Investigating numerical correctness failures, such as NaN/Inf outputs or mismatches in an existing FlyDSL kernel. (see Validation--Agent loads skills correctly--case2)
- Fixing compilation or frontend tracing failures in an existing FlyDSL kernel.
First of all, design a kernel with compile failure, e.g.
image


```
geak \
  --repo "$REPO" \
  --kernel-url "$REPO/tracing_failure_range_constexpr_kernel.py" \
  --task "This is an existing FlyDSL kernel that fails with a compilation/frontend error in a fixed-width loop inside @flyc.kernel. First reproduce the failure from the provided test command, then identify the root cause and make the minimum code change needed so the same entry point passes correctness. Preserve the current operator, launch path, and validation behavior. Do not optimize performance or rewrite the kernel into a different form." \
  --test-command "FLYDSL_RUNTIME_ENABLE_CACHE=0 python tracing_failure_range_constexpr_kernel.py --seed 1234" \
  --yolo
```
image

**Comment by amd-danli103** [comment]
### How skills benefit the tasks
#### Case1. writing a new flyDSL softmax kernel
with exactly the same prompt
```
geak --repo "$REPO" --kernel-url "$REPO/base_softmax.py" --task "用 FlyDSL 写一个 first-correct 的 softmax kernel，给最小 correctness harness，先不用做性能优化。" --yolo
```
- Without `flydsl-tile-programming/SKILL.md` and `use_skill=False`
1.  new kernel development done with extra exploratory test files included, e.g. `base_softmax.py` plus `test_f64_softmax.py` and `test_flydsl_softmax.py`
- With `flydsl-tile-programming/SKILL.md` and `use_skill=True`
from previous log https://github.com/AMD-AGI/GEAK/pull/175#issuecomment-4303854789, we know that
1. the correct skill `flydsl-tile-programming` has been loaded.
2. Process-wise, **the agent follows the intended tile-programming path**: reduction classification, FlyDSL kernel-shape work, harness-stable single-file replacement.
3.  new kernel development done with minimal target-file rewrite, producing a cleaner and more minimal final artifact. Its winning patch is limited to only `base_softmax.py`. This supports a moderate claim that the skill improves artifact quality by steering the agent toward a more first-correct and structurally tidy solution.

#### Case2. the buggy flyDSL softmax kernel which produces NaN
with exactly the same prompt
```
geak --repo "$REPO" \
  --kernel-url "$REPO/buggy_softmax_kernel.py" \
  --test-command "python repro_buggy_softmax_nan.py" \
  --task "这是一个已有的FlyDSL softmax kernel。先运行 repro_buggy_softmax_nan.py 复现 NaN/inf 和 correctness mismatch，再定位 root cause 并修复。保持现有 kernel 入口不变，先不要做性能优化，也不要整体重写。" \
  --yolo
```
- Without `flydsl-debug-kernel/SKILL.md` and `use_skill=False`
1. the NaN issue fixed using **13 steps / 137s time spent for sub-agent**
The no-skill run briefly pursued a secondary bug hypothesis that ultimately proved unnecessary for the final fix (refer to `~/GEAK/optimization_logs/softmax_20260427_074557/results/round_1/parallel_0/task_0.log`)
```
THOUGHT: Now I'm confident about the bugs. Let me trace through the logic:
**Bug 1 (line 75)**: `c_pad_sentinel = float("inf")` (positive infinity)
...
**Fix for Bug 1**: Change `float("inf")` to `float("-inf")` on line 75.
**Bug 2 (line 176)**: `new_buffer.append((exp_val, is_valid))` stores the raw `exp_val` instead of `safe_exp`
...
Actually, let me reconsider.
...
So **Bug 1 is the primary root cause**. Bug 2 is a code quality issue but doesn't cause incorrect results after Bug 1 is fixed.
Let me fix Bug 1:
```
- With `flydsl-debug-kernel/SKILL.md` and `use_skill=True`
from previous log https://github.com/AMD-AGI/GEAK/pull/175#issuecomment-4303854789, we know that
1. the correct skill `flydsl-debug-kernel` has been loaded.
5. Process-wise, **the skills-on run adhered more closely to the workflow prescribed by the debug skill**: it began by replaying the provided repro under cache bypass, then classified the failure, isolated the root cause, applied a minimal fix, re-ran the repro, and only afterwards expanded verification to additional shapes and the final \save_and_test` step. 
6. the NaN issue fixed using **11 steps / 100s time spent for sub-agent**, debug is more efficiently and time-saving!

**Comment by amd-danli103** [comment]
Hi @amd-ethany and @johayang-amd, I've restructured the FlyDSL skills based on our previous discussion. Following the `pytorch2flydsl-translation/` pattern, I've consolidated the three separate FlyDSL skills (`flydsl-optimization`, `flydsl-debug-kernel`, `flydsl-tile-programming`) into a single unified `skills/flydsl/` folder:

```
skills/flydsl/
  SKILL.md                          ← unified summary (write / optimize / debug workflows)
  docs/
    flydsl_optimization.md          ← optimization strategies (moved from flydsl-optimization/)
    flydsl_debug_kernel.md          ← correctness debugging guide
    flydsl_tile_programming.md      ← tile programming guide
```

Key changes:
- **Single skill entry point**: One `SKILL.md` summarizing all three workflows with a task-based routing table
- **Detailed docs in `docs/`**: Each skill area is one `.md` file, referenced from the summary
- **No content changes**: The technical content of all three skills is preserved as-is; only the structure and YAML frontmatter were updated
- **`flydsl-optimization/` absorbed**: The existing optimization skill is moved into the unified structure

Could you please take another look? or pls feel free to invite other related forks. Thanks!

**Review by yueliu14 — APPROVED** [review]

**Review by amd-ethany — APPROVED** [review]
